aten/src/ATen/FunctionalStorageImpl.cpp

```diff
@@ -0,0 +1,110 @@
+#include <ATen/FunctionalStorageImpl.h>
+
+#include <ATen/FunctionalTensorWrapper.h>
+#include <ATen/core/LegacyTypeDispatch.h>
+#include <c10/util/Exception.h>
+#include <vector>
+
+namespace at {
+namespace functionalization {
+
+Alias::Alias(const at::Tensor& base) {
+  TORCH_INTERNAL_ASSERT(!at::functionalization::impl::isFunctionalTensor(base));
+  base_ = base;
+}
+
+const at::Tensor& Alias::base() const {
+  return base_;
+}
+
+// metas is taken by value on purpose - we want to copy the vector.
+void Alias::add_update(const at::Tensor& updated_val, std::vector<ViewMeta> metas) {
+  updates_.push_back({updated_val, metas});
+  generation_++;
+}
+
+// Note [Functionalization: Alias Removal Part 2]
+// See Note [Functionalization: Alias Removal] for more details.
+// This function applies a single update from one of the views to the Alias object.
+// We start out with <original_base> and <mutated_view>, and our goal is to end up with <mutated_base>.
+void Alias::apply_update(const Update& update) {
+  at::AutoDispatchBelowFunctionalize guard;
+  at::Tensor t = update.new_val;
```

Comment:
Naively, the tensors here NOT being functional tensors seems to suggest the below functionalize guard is not strictly necessary. So it would be good to have an explanation what it is actually helpful for.

Reply:
I had to double-check why I needed this. You're right that I don't think we'll need it for other backends, but (I think) we need it for functorch. I'll add a comment.

That's because functorch's DynamicLayer logic uses TLS to make sure sure that `DynamicLayerFrontFallback` goes to the right kernel; so at this point in the code, `Functionalize` is in the TLS include set.

I could imagine that logic maybe being re-written to not add any of those transform keys into TLS (and instead just reading off of the `key_set()` from each input wrapper argument). But I'm not sure if there are any subtleties that prevent doing that.